File: framework/languages/dotnet/e2e/YieldDispatch/Server/Play/PlayHostFactory.cs
using Systems.Zlink;
using YieldDispatch.Server.Play.Handlers;
using YieldDispatch.Server.Play.Spots;
using YieldDispatch.Shared;
using Zlink.Framework.AspNetCore;
using Zlink.Framework.Contracts.Dispatch;

namespace YieldDispatch.Server.Play;

internal static class PlayHostFactory
{
    public static WebApplication Create(string[] args)
    {
        var options = PlayOptions.Parse(args);
        Directory.CreateDirectory(options.LogDir);

        var builder = WebApplication.CreateBuilder(args);
        builder.Logging.ClearProviders();
        builder.Logging.AddSimpleConsole(console =>
        {
            console.SingleLine = true;
            console.TimestampFormat = "HH:mm:ss.fff ";
        });
        builder.WebHost.UseUrls(options.HttpUrl);
        builder.Services.AddSingleton(new EvidenceStore(options.Rid, options.EvidenceFile));
        builder.Services.AddSingleton(new NodeOptions(options.Rid));
        builder.Services.AddZLinkFramework(framework =>
        {
            framework.AddHandlersFromAssemblyOf(typeof(Program));
            framework.ConfigureDispatch()
                .MessageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
                .TraceLogFile(Path.Combine(options.LogDir, $"{options.Rid}-flow.log"))
                .TraceLabel(options.Rid);
            framework.AddRouteMesh(YieldDispatchNames.ControlChannel)
                .EnableServer(options.ControlEndpoint)
                .EnableClient()
                .SetRoutingId(RoutingId.From(options.Rid))
                .AddRequestHandler<BindYieldActorsControlHandler, BindYieldActorsReq, BindYieldActorsRes>(
                    "BindYieldActorsReq")
                .AddRequestHandler<EnsureSpotControlHandler, EnsureSpotReq, EnsureSpotRes>("EnsureSpotReq")
                .AddRequestHandler<YieldEvidenceControlHandler, YieldEvidenceReq, YieldEvidenceRes>(
                    "YieldEvidenceReq")
                .AddRequestHandler<YieldEvidenceWaitControlHandler, YieldEvidenceWaitReq, YieldEvidenceRes>(
                    "YieldEvidenceWaitReq");
            framework.AddClientServerChannel(YieldDispatchNames.DelayChannel)
                .EnableClient(options.DelayEndpoint)
                .SetRoutingId(RoutingId.From(options.Rid));
            framework.AddRouteMesh(YieldDispatchNames.SpotRouteChannel)
                .EnableServer(options.SpotRouteEndpoint)
                .EnableClient()
                .SetRoutingId(RoutingId.From(options.Rid));
            framework.AddSpotMesh(YieldDispatchNames.SpotChannel)
                                .EnableRouter(options.SpotRouterEndpoint)
                .SetRoutingId(RoutingId.From(options.Rid))
                .EnablePubSub(options.SpotPubEndpoint)
                .AddEntrySpot<YieldEntrySpot>()
                .AddActorFactory<YieldActorFactory>(YieldDispatchNames.ActorType)
                .AddSpotFactory<YieldProbeSpot>();
        });

        var app = builder.Build();
        app.MapGet("/health", () => Results.Ok(new { status = "ready", role = "play", options.Rid }));
        app.MapGet("/evidence", (EvidenceStore evidence) => Results.Ok(evidence.Snapshot()));
        return app;
    }
}

File: framework/languages/node/e2e/YieldDispatch/Server/Play/play-host-factory.ts
import fs from 'node:fs';
import { Module } from '@nestjs/common';
import { NestFactory } from '@nestjs/core';
import { ZLinkMessageFlowLogMode } from '@zlink-systems/framework';
import { ZLinkModule, zlinkFramework } from '@zlink-systems/nestjs';
import { YieldDispatchNames } from '../../Shared/messages';
import { EvidenceStore } from '../Support/evidence-store';
import { closeHttpServer, startHttpServer } from '../Support/http-server';
import { parsePlayOptions } from './Configuration/play-options';
import { HoldCommandHandler, ProbeCommandHandler, WorkerYieldCommandHandler, YieldCommandHandler, YieldRequestHandler } from './Handlers/basic-spot-handlers';
import {
  BindYieldActorsControlHandler,
  EnsureSpotControlHandler,
  YieldEvidenceControlHandler,
  YieldEvidenceWaitControlHandler
} from './Handlers/control-handlers';
import { YieldCancelCommandHandler, YieldTimeoutCommandHandler } from './Handlers/failure-spot-handlers';
import { RemoteSpotYieldCommandHandler, RemoteSpotYieldHandler } from './Handlers/remote-spot-handlers';
import { TimerStartCommandHandler, TimerStopCommandHandler, YieldTimerHandler } from './Handlers/timer-spot-handlers';
import {
  EntryActorFastHandler,
  EntryActorFastSendHandler,
  EntryActorJoinYieldHandler,
  EntryActorPushYieldHandler,
  EntryActorYieldHandler,
  SpotActorFastHandler,
  SpotActorFastSendHandler,
  SpotActorPushYieldHandler,
  SpotActorYieldHandler,
  YieldActorFactory,
  YieldEntrySpot
} from './Spots/yield-actors';
import { YieldProbeSpot } from './Spots/yield-probe-spot';

export async function startPlayHost(args: readonly string[]): Promise<void> {
  const options = parsePlayOptions(args);
  fs.mkdirSync(options.logDir, { recursive: true });
  const evidence = new EvidenceStore(options.rid, options.evidenceFile);
  let stopping = false;

  class PlayModule {}
  Module({
    imports: [
      ZLinkModule.forRootFactory({
        useFactory: () => {
          const builder = zlinkFramework();
          builder
            .options({
              registrySpotRemoteAddresses: {
                namespace: YieldDispatchNames.spotChannel,
                routerChannelId: YieldDispatchNames.spotRouteChannel
              }
            })
            .configureDispatch()
              .messageFlow(ZLinkMessageFlowLogMode.KeyTransitions)
              .traceLogFile(`${options.logDir}/${options.rid}-flow.log`)
              .traceLabel(options.rid);
          builder.useDiscovery().addRegistryEndpoint(options.registryRouterEndpoint);
          builder.addRouteMesh(YieldDispatchNames.controlChannel)
            .enableRouter(options.controlEndpoint)
            .routingId(options.rid)
            .addRequestHandler('EnsureSpotReq', EnsureSpotControlHandler)
            .addRequestHandler('BindYieldActorsReq', BindYieldActorsControlHandler)
            .addRequestHandler('YieldEvidenceReq', YieldEvidenceControlHandler)
            .addRequestHandler('YieldEvidenceWaitReq', YieldEvidenceWaitControlHandler);
          builder.addClientServerChannel(YieldDispatchNames.delayChannel)
            .enableClient(options.delayEndpoint);
          builder.addRouteMesh(YieldDispatchNames.spotRouteChannel)
            .enableRouter(options.spotRouteEndpoint)
            .routingId(options.rid);
          builder.addSpotMesh(YieldDispatchNames.spotChannel)
            .routingId(options.rid)
            .enableRouter(options.spotRouterEndpoint)
            .enablePubSub(options.spotPubEndpoint)
            .addEntrySpot(YieldEntrySpot)
            .actorFactory(YieldDispatchNames.actorType, YieldActorFactory)
            .addSpotFactory(YieldProbeSpot);
          return builder.build();
        }
      })
    ],
    providers: [
      { provide: EvidenceStore, useValue: evidence },
      EnsureSpotControlHandler,
      BindYieldActorsControlHandler,
      YieldEvidenceControlHandler,
      YieldEvidenceWaitControlHandler,
      HoldCommandHandler,
      YieldCommandHandler,
      YieldRequestHandler,
      WorkerYieldCommandHandler,
      YieldTimeoutCommandHandler,
      YieldCancelCommandHandler,
      ProbeCommandHandler,
      RemoteSpotYieldHandler,
      RemoteSpotYieldCommandHandler,
      TimerStartCommandHandler,
      TimerStopCommandHandler,
      YieldTimerHandler,
      YieldActorFactory,
      YieldEntrySpot,
      EntryActorYieldHandler,
      EntryActorFastHandler,
      EntryActorFastSendHandler,
      EntryActorJoinYieldHandler,
      EntryActorPushYieldHandler,
      SpotActorYieldHandler,
      SpotActorFastHandler,
      SpotActorFastSendHandler,
      SpotActorPushYieldHandler,
      YieldProbeSpot
    ]
  })(PlayModule);

  const app = await NestFactory.createApplicationContext(PlayModule, { logger: false, abortOnError: false });
  const server = await startHttpServer(options.httpUrl, [
    { method: 'GET', path: '/health', handle: () => ({ status: 'ready', role: 'play', rid: options.rid }) },
    { method: 'GET', path: '/evidence', handle: () => evidence.snapshot() },
    {
      method: 'POST',
      path: '/shutdown',
      handle: () => {
        stopping = true;
        return { status: 'stopping' };
      }
    }
  ]);
  while (!stopping) {
    await new Promise((resolve) => setTimeout(resolve, 100));
  }
  await closeHttpServer(server);
  await app.close();
}
